src/pluginextension.h

```diff
@@ -0,0 +1,191 @@
+	 * no meaningful suggestions.
+	 * @param data User data passed during the @c plugin_extension_register()
+	 * call.
+	 *
+	 * @since 2.1
+	 **/
+	void (*autocomplete_perform)(GeanyDocument *doc, gboolean force, gpointer data);
+
+	/**
+	 * Pointer to function called by Geany to check whether the plugin
+	 * implements calltips containing function signatures for the provided
+	 * document.
+	 *
+	 * @see @c autocomplete_provided() for more details.
+	 *
+	 * @since 2.1
+	 **/
+	gboolean (*calltips_provided)(GeanyDocument *doc, gpointer data);
+	/**
+	 * Pointer to function called by Geany to inform the plugin to show calltips.
+	 *
+	 * @see @c autocomplete_perform() for more details.
+	 *
+	 * @since 2.1
+	 **/
+	void (*calltips_show)(GeanyDocument *doc, gboolean force, gpointer data);
+
+	/**
+	 * Pointer to function called by Geany to check whether the plugin implements
+	 * symbol definition/declaration goto functionality.
+	 *
+	 * @see @c autocomplete_provided() for more details.
+	 *
+	 * @since 2.1
+	 **/
+	gboolean (*goto_provided)(GeanyDocument *doc, gpointer data);
+	/**
+	 * Pointer to function called by Geany to inform the plugin to perform
+	 * symbol goto.
+	 *
+	 * @param doc The document for which symbol goto is being performed.
+	 * @param pos Scintilla position in the document at which the goto request
+	 * was invoked (typically corresponds to the caret position or the position
+	 * where the corresponding mouse event happened).
+	 * @param definition If @c TRUE, this is the go to definition request, if
+	 * @c FALSE, this is the goto declaration request.
+	 * @param data User data passed during the @c plugin_extension_register()
+	 * call.
+	 *
+	 * @return Plugins should return @c TRUE if the goto was performed, @c FALSE
+	 * otherwise. Plugins that work asynchronously and do not know the result
+	 * at the moment this function is called should return @c TRUE. However, such
+	 * plugins can perform some synchronous pre-check such as whether there is
+	 * an identifier at the caret position and if the necessary conditions are
+	 * not satisfied and the plugins know that goto cannot be performed, they
+	 * should return @c FALSE.
+	 *
+	 * @since 2.1
+	 **/
+	gboolean (*goto_perform)(GeanyDocument *doc, gint pos, gboolean definition, gpointer data);
+
+	/**
+	 * Pointer to function called by Geany to check whether the plugin implements
+	 * additional symbol (e.g. type) highlighting in Scintilla.
+	 *
+	 * @see @c autocomplete_provided() for more details.
+	 * @note There is no function in the @c PluginExtension structure informing
+	 * plugins to perform symbol highlighting. Plugins
+	 * implementing symbol highlighting should perform it at the appropriate
+	 * moments based on Scintilla and Geany events such as when the document
+	 * becomes visible or when the document is modified.
+	 *
+	 * @since 2.1
+	 **/
+	gboolean (*symbol_highlight_provided)(GeanyDocument *doc, gpointer data);
```

Comment:
IMO this is a recipe for getting it wrong. Why doesn't Geany define "appropriate moments based on Scintilla and Geany events" and calls a `symbol_highlight_perform()` then?

Reply:
IIUC the issue is that it basically depends on when the symbols get ready, which is not something Geany can control for e.g. LSP. Sure, it could be added to various places, but for the case of the LSP plugin it would *never* be useful.  For other synchronous applications it could potentially be useful I guess, so possibly calling it where Geany already does might be a reasonable thing, and an async plugin would just do nothing there… dunno.

Reply:
> IMO this is a recipe for getting it wrong. Why doesn't Geany define "appropriate moments based on Scintilla and Geany events" and calls a symbol_highlight_perform() then?

@kugel- I did it this way originally but there were situations where it didn't work. Consider for instance LSP server start - once it is started, it should colorize the current document. But for Geany "nothing happened" in the meantime - there was no Geany event that would trigger the colorization as Geany knows nothing about LSP processes and their start. So in this case you wouldn't get any colorization and documents would only get colorized once you trigger some Geany event like when you e.g. type something or switch tabs.

There will be a similar problem with the symbol tree update in https://github.com/geany/geany/pull/3910 where I think it would be best to have some `symbol_tree_reload()` function the plugin could call to force Geany reload the symbol tree as Geany doesn't know about all the situations in which it may be necessary.

To be precise, all of the other calls from this API suffer from the same problem - it's just that this isn't critical for something like autocompletion, calltips, or symbol goto which are originally processed by Geany before the server is started and only then switch to the server implementation.

Reply:
Hmm perhaps we should find a generic solution for the LSP support to do some stuff asynchronously (not necessarily now, but gradually and in the meantime we define the LSP API "not stable yet").

For example, for colorization, the  LSP plugin can signal readyness to Geany's mainloop, and when Geany handles that "ready event", it would finally call `symbol_highlight_perform()`. This could even be used to complete Geany's own highlighting as  well.

I find it highly problematic that it's up to the LSP plugins to get the places, from where highlighting is triggered, right. At best the author looks where Geany calls `document_highlight_tags()` and draw conclusions from there. However, he may miss calls and/or it may even change between Geany versions and the plugin will regress in one version or the other.

Reply:
> I find it highly problematic that it's up to the LSP plugins to get the places, from where highlighting is triggered, right. At best the author looks where Geany calls document_highlight_tags() and draw conclusions from there. 

Nothing like that is necessary - the plugin just invokes it based on whether it determines the highlighting is necessary which is in 2 situations which are quite obvious from the plugin writer's perspective:
- contents of the document changes, i.e. in `SC_MOD_INSERTTEXT | SC_MOD_DELETETEXT` scintilla notification
- a tab is switched - i.e. `document-activate` Geany signal

That's it (and I actually don't know when exactly Geany triggers it and don't really care). I think the logic when to do it is pretty clear and much simpler than having to forward it to Geany.

There's another related problem - Geany uses Scintilla's lexer keyword index

https://github.com/geany/geany/blob/9bf5769f582cbf3d82a1faca035e7f63e8908afe/src/document.c#L2722

and highlights those by setting the keywords using `sci_set_keywords()`. I pretty much copied this part and it's one type of highlighting. The trouble with this approach is:
1. It's limited to only those lexers that support it
2. It highlights all the occurrences of `foo` (if `foo` is defined as a type somewhere) no matter whether `foo` is actually used as a type in the particular file

This is why the colorization using `INDIC_TEXTFORE` was added in https://github.com/techee/geany-lsp/issues/18 based on @elextr 's suggestion. This allows colorization of only specific positions in the document so when you have `foo` as a type and `foo` as a variable in the same document, the first `foo` gets colorized and the second one doesn't. (One limitation of this approach is that the text cannot be made bold, see the linked issue above for more discussion.)

This second mode wouldn't be possible with current Geany's colorization.

Reply:
It makes me sad that significant improvements like that are only available on your lsp plugin. This suggests lots of future new  hotness only hits one or the other LSP plugin and Geany's general functionality is left behind. This (colorization of keywords vs types) has always been a (minor) annoyance and I'm wondering why Geany can't do this right (I always assumed that it's a strict Scintilla limitation)?

Reply:
Its not really a Lexilla issue, lexers do not have any idea about types so they cannot distinguish which instances of a name are types and which are variables and which are something else.

The problem is that ctags doesn't know either, it only parses declarations of names, not their uses.  It does not parse most code to the point of tracing visibility and determining which declaration of a name is visible at the point of use, so its cannot provide information to colour different instances of a name differently.

Its not Geany capability that is "left behind", its the external tools it uses, Lexilla and ctags, neither of which is designed to do what is required to provide the capability.  All LSPs are is another external tool that does provide the capability and more as well, and so Geany is being extended to make use of it, just like it makes use of ctags parsers or Lexilla lexers.  Its nothing new, Geany has always depended on external tools, LSPs are just another one.

Reply:
> This (colorization of keywords vs types) has always been a (minor) annoyance and I'm wondering why Geany can't do this right (I always assumed that it's a strict Scintilla limitation)?

Pretty much what @elextr said - type names in Geany are obtained from ctags and then added to Scintilla as "keywords" so they get colorized but ctags cannot (and never will) provide the information where the given type is used so we won't be able to colorize the precise instances, just all of them. On the other hand, you get these from the LSP server from the "semantic token" call with exact ranges so these can then be colorized precisely (when `INDIC_TEXTFORE` colorization is enabled).

Reply:
> This (colorization of keywords vs types) has always been a (minor) annoyance

And my 2c, thats a personal thing, for me its minor for C, where the same name cannot be a type and a function and a variable, but its not minor for me with C++ where type names and variable names and function names can overlap, and often do because the standard library took all the obvious ones (like "get" :-).